proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -7040,10 +7040,13 @@ method merge_refills_simple
      , (clarsimp simp: vs_all_heap_simps obj_at_kh_kheap_simps pred_map_simps refills_sum_def)?)
 
 lemma merge_refills_refills_sum:
-  "\<lbrace>\<lambda>s. pred_map (\<lambda>cfg. refills_sum (scrc_refills cfg) = scrc_budget cfg)
-                 (sc_refill_cfgs_of s) sc_ptr
-        \<and> pred_map (\<lambda>cfg. 1 < length (scrc_refills cfg)) (sc_refill_cfgs_of s) sc_ptr\<rbrace>
-   merge_refills sc_ptr
+  "\<lbrace>\<lambda>s. if sc_ptr' = sc_ptr
+        then pred_map (\<lambda>cfg. refills_sum (scrc_refills cfg) = scrc_budget cfg)
+                      (sc_refill_cfgs_of s) sc_ptr
+             \<and> pred_map (\<lambda>cfg. 1 < length (scrc_refills cfg)) (sc_refill_cfgs_of s) sc_ptr
+        else pred_map (\<lambda>cfg. refills_sum (scrc_refills cfg) = scrc_budget cfg)
+                      (sc_refill_cfgs_of s) sc_ptr\<rbrace>
+   merge_refills sc_ptr'
```

Comment:
We’d want to refactor the common clause out of if-then-else here. And then having an implication `sc_ptr’ = sc_ptr --> ...` as the second precondition would be good.